### PR TITLE
raft: port etcd(set leader id in stepfollower)

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
raft: set leader id in stepfollower

### CODE REVIEW

**Comment by ngaut** [comment]
Could you add the original reference link ?

File: tests/test_raft.rs [hunk 1]
```diff
@@ -1693,6 +1712,10 @@ fn test_restore_from_snap_msg() {
 
     sm.step(m).expect("");
 
+    if sm.leader_id != 1 {
```
Comment (BusyJay) [inline tests/test_raft.rs]: use `assert_eq` instead.

File: tests/test_raft.rs [hunk 2]
```diff
@@ -698,7 +698,26 @@ fn test_dueling_candidates() {
     nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
     nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
 
+    {
+        // 1 becomes leader since it receives votes from 1 and 2
+        let sm = nt.peers.get(&1).unwrap();
+        if sm.state != StateRole::Leader {
```
Comment (BusyJay) [inline tests/test_raft.rs]: `nt.peers[&1].state`

**Comment by siddontang** [comment]
LGTM

**Comment by BusyJay** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
https://github.com/coreos/etcd/pull/5388
https://github.com/coreos/etcd/pull/5919

**Comment by siddontang** [comment]
Can we merge it? @zhangjinpeng1987

**Pull request merged**